torch/distributed/tensor/_api.py

```diff
@@ -764,6 +764,51 @@ def distribute_tensor(
     )
 
 
+def _shard_tensor(
+    full_tensor: torch.Tensor,
+    placements: Sequence[Shard],
+    device_mesh: Optional[DeviceMesh] = None,
+) -> "DTensor":
+    """
+    Locally shards a full tensor based on indicated sharding arrangement, and
+    returns a DTensor containing the local shard.
+
+    .. warning:: This is a private API purposed to skip the communication
+        otherwise required by `distribute_tensor`. It is only applicable to a
+        case where all ranks have the same `full_tensor`. This API will not
+        check for data equality between ranks, it is thus user's responsibility
+        to ensure the `full_tensor` is the same across ranks.
+
+    Args:
+        full_tensor (torch.Tensor): the full tensor to be sharded.
+        placements (Sequence[:class:`Shard`]): the placements that
+            describes how to place the local tensor on DeviceMesh.
+        device_mesh (:class:`DeviceMesh`, optional): DeviceMesh to place the
+            DTensor.  Must have same dimension as the number of placements.
+            If not specified, would be retrieve from current context.
+
+    Returns:
+        A :class:`DTensor` object with the shard as its local tensor.
+
+    Examples:
+        >>> # xdoctest: +SKIP("need world_size and rank")
+        >>> device_mesh = dist.init_device_mesh("cuda", (world_size,))
+        >>> full_tensor = torch.arange(world_size, device=f"cuda:{rank}")
+        >>> dtensor = _shard_tensor(full_tensor, [Shard(1)], device_mesh)
+    """
+    device_mesh = device_mesh or _mesh_resources.get_current_mesh()
+
+    shape, offset = compute_local_shape_and_global_offset(
+        full_tensor.shape, device_mesh, placements
+    )
+    slices = [
+        slice(cur_offset, cur_offset + cur_shape)
+        for cur_shape, cur_offset in zip(shape, offset)
+    ]
+    local_tensor = full_tensor[slices]
+    return DTensor.from_local(local_tensor, device_mesh, placements)
```

Comment:
You will need to pass the `shape` and `stride` for uneven tensor for `from_local`. Otherwise, the shape and stride would be inferred from the local tensor as it is uniformly distributed. See example: https://github.com/pytorch/pytorch/blob/main/torch/distributed/_state_dict_utils.py#L566-L572

Reply:
Thanks for the heads-up. I don't have the shape or stride info, and this function assumes things are the same across ranks, so the infer would be fine.